frontend/src/core/cells/data-source-connections.ts

```diff
@@ -32,16 +31,11 @@ const {
     state: DataSourceState,
     opts: { connections: DataSourceConnection[] },
   ): DataSourceState => {
-    const newConnections = opts.connections;
-
-    for (const newConn of newConnections) {
-      const existingConnection = state.connectionsMap.get(newConn.name);
-      if (existingConnection) {
-        // update the existing connection
-        state.connectionsMap.set(existingConnection.name, newConn);
-      } else {
-        state.connectionsMap.set(newConn.name, newConn);
-      }
+    // update existing connections with latest values
+    // add new ones if they don't exist
+    // Backend will dedupe by connection name & keep the latest, so we use this as the key
+    for (const conn of opts.connections) {
+      state.connectionsMap.set(conn.name, conn);
```

Comment:
because this is used in jotai and react, we need to create treat state as immutable, otherwise react-compiler and other optimizations will prevent re-renders. 

1. we should re-create the map at the top:
```ts
const connectionsMap = new Map(state. connectionsMap)
// do things
return {...state, connectionsMap };
```

2. we can actually improve our types (and I likely haven't set a good example in the codebase), but we can update:

```diff
export interface DataSourceState {
-  connectionsMap: Map<ConnectionName, DataSourceConnection>;
+  connectionsMap: ReadonlyMap<ConnectionName, DataSourceConnection>;
}
```

which will force you to do `new Map()` to mutate it